File: packages/python/fayda_sdk/userinfo/fetch.py
from typing import Union

import httpx
from jose import jwt

from fayda_sdk.auth import AuthModule
from fayda_sdk.config import FaydaConfig
from fayda_sdk.errors import FaydaSandboxError
from fayda_sdk.models import FaydaUser
from fayda_sdk.token import TokenModule


def _mock_user() -> FaydaUser:
    return FaydaUser(
        sub="mock-fayda-id-000001",
        name="Abebe Bikila",
        name_am="አበበ ቢኪላ",
        name_en="Abebe Bikila",
        email="abebe.bikila@example.com",
        phone_number="+251911000000",
        picture="https://placehold.co/150x150?text=AB",
        gender="male",
        birthdate="1932-08-07",
        address={
            "formatted": "Addis Ababa, Ethiopia",
            "locality": "Addis Ababa",
            "country": "ET",
        },
    )


class UserInfoModule:
    def __init__(
        self,
        config: FaydaConfig,
        http_client: Union[httpx.Client, httpx.AsyncClient],
        auth_module: AuthModule,
        token_module: TokenModule,
    ):
        self.config = config
        self.http_client = http_client
        self.auth_module = auth_module
        self.token_module = token_module

    def get(self, access_token: str) -> FaydaUser:
        if self.config.sandbox:
            return _mock_user()

        response = self.http_client.get(
            self.config.userinfo_endpoint,
            headers={"Authorization": f"Bearer {access_token}"},
        )

        if response.status_code != 200:
            raise Exception(f"UserInfo request failed: {response.status_code}")

        jwt_text = response.text
        claims = jwt.get_unverified_claims(jwt_text)
        return FaydaUser.from_claims(claims)

    def get_from_code(self, code: str, state: str, expected_state: str, code_verifier: str) -> FaydaUser:
        self.auth_module.validate_callback(state, expected_state)
        tokens = self.token_module.exchange(code, code_verifier)
        return self.get(tokens.access_token)

    def get_mock(self) -> FaydaUser:
        if not self.config.sandbox:
            raise FaydaSandboxError("get_mock() is only available in sandbox mode")
        return _mock_user()

File: packages/node/src/userinfo/fetch.ts
import { decodeJwt } from "jose";
import { ResolvedConfig } from "../config.js";
import { FaydaUser, fromUserInfoClaims } from "../models.js";
import { FaydaUserInfoError, FaydaSandboxError } from "../errors/index.js";
import { AuthFlow } from "../auth/flow.js";
import { TokenExchange } from "../token/exchange.js";

const MOCK_USER: FaydaUser = {
  sub: "mock-fayda-id-000001",
  name: "Abebe Bikila",
  nameAm: "አበበ ቢኪላ",
  nameEn: "Abebe Bikila",
  email: "abebe.bikila@example.com",
  phoneNumber: "+251911000000",
  picture: "https://placehold.co/150x150?text=AB",
  gender: "male",
  birthdate: "1932-08-07",
  address: {
    formatted: "Addis Ababa, Ethiopia",
    locality: "Addis Ababa",
    country: "ET",
  },
};

export class UserInfoFetcher {
  constructor(
    private config: ResolvedConfig,
    private authFlow: AuthFlow,
    private tokenExchange: TokenExchange
  ) {}

  async get(accessToken: string): Promise<FaydaUser> {
    if (this.config.sandbox) {
      return MOCK_USER;
    }

    const response = await fetch(this.config.userInfoEndpoint, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });

    if (!response.ok) {
      throw new FaydaUserInfoError(
        `UserInfo request failed: ${response.status}`
      );
    }

    const jwtText = await response.text();
    const claims = decodeJwt(jwtText) as Record<string, unknown>;
    return fromUserInfoClaims(claims);
  }

  async getFromCode(params: {
    code: string;
    state: string;
    expectedState: string;
    codeVerifier: string;
  }): Promise<FaydaUser> {
    this.authFlow.validateCallback(params.state, params.expectedState);
    const tokens = await this.tokenExchange.exchange(
      params.code,
      params.codeVerifier
    );
    return this.get(tokens.accessToken);
  }

  getMock(): FaydaUser {
    if (!this.config.sandbox) {
      throw new FaydaSandboxError("getMock() is only available in sandbox mode");
    }
    return MOCK_USER;
  }
}
